include/gc/Dialect/OneDNNGraph/OneDNNGraphOps.td

```diff
@@ -24,22 +24,38 @@ include "gc/Dialect/OneDNNGraph/OneDNNGraphTypes.td"
 class OneDNNGraph_Op<string mnemonic, list<Trait> traits = []> :
         Op<OneDNNGraphDialect, mnemonic, traits>;
 
+class OneDNNGraph_ElemwiseUnaryOp<string mnemonic, list<Trait> traits = []> :
+    OneDNNGraph_Op<mnemonic, traits # 
+    [SameOperandsAndResultType]> {
+  let arguments = (ins OneDNNGraph_FloatTensor:$operand);
+  let results = (outs OneDNNGraph_FloatTensor:$result);
+
+  let assemblyFormat =
+      "operands attr-dict `:` functional-type(operands, results)";
+}
+
 class OneDNNGraph_ElemwiseBinaryOp<string mnemonic, list<Trait> traits = []> :
-    OneDNNGraph_Op<mnemonic, traits # [SameOperandsAndResultElementType, InferTensorType, 
-    ResultsBroadcastableShape]> {
-  let arguments = (ins OneDNNGraph_FloatTensor:$input_0, 
-                       OneDNNGraph_FloatTensor:$input_1);
+    OneDNNGraph_Op<mnemonic, traits # 
+    [SameOperandsAndResultElementType, InferTensorTypeAdaptor, ResultsBroadcastableShape]> {
+  let arguments = (ins OneDNNGraph_FloatTensor:$input_a, 
+                       OneDNNGraph_FloatTensor:$input_b,
+                       DefaultValuedOptionalAttr<BoolAttr, "true">:$auto_broadcast);
   let results = (outs OneDNNGraph_FloatTensor:$result);
 
   let assemblyFormat =
       "operands attr-dict `:` functional-type(operands, results)";
 }
 
-class OneDNNGraph_ElemwiseUnaryOp<string mnemonic, list<Trait> traits = []> :
-    OneDNNGraph_Op<mnemonic, traits # [SameOperandsAndResultType]> {
-  let arguments = (ins OneDNNGraph_FloatTensor:$operand);
+class OneDNNGraph_ReduceOp<string mnemonic, list<Trait> traits = []> :
+    OneDNNGraph_Op<mnemonic, traits # 
+    [SameOperandsAndResultElementType, InferTensorTypeAdaptor]> {
+  let arguments = (ins OneDNNGraph_FloatTensor:$operand,
+                       DefaultValuedOptionalAttr<DenseI64ArrayAttr, "{}">:$axes,
+                       DefaultValuedOptionalAttr<BoolAttr, "false">:$keep_dims);
```

Comment:
If attr `keep_dims=true`, we will lower the reduce op to `linalg.reduce` as below?
```
%reduce = linalg.reduce
      ins(%input:tensor<16x32x64xf32>)
      outs(%init:tensor<16x1x64xf32>)
      dimensions = [1]
      (%in: f32, %out: f32) {
        %0 = arith.addf %out, %in: f32
        linalg.yield %0: f32
      }
```

Reply:
> LGTM. I have a question: If attr `keep_dims=true`, we will lower the reduce op to `linalg.reduce` as below?
> 
> ```
> %reduce = linalg.reduce
>       ins(%input:tensor<16x32x64xf32>)
>       outs(%init:tensor<16x1x64xf32>)
>       dimensions = [1]
>       (%in: f32, %out: f32) {
>         %0 = arith.addf %out, %in: f32
>         linalg.yield %0: f32
>       }
> ```

No, `linalg.reduce` only allows `keep_dims == False`, so an extra `tensor.expand_shape` will be inserted.

Reply:
Yeah, I just found that linalg.reduce don't allow this attr to be true. It should be:
```
%0 = tensor.empty() : tensor<16x64xf32>
  %reduce = linalg.reduce
      ins(%arg0:tensor<16x32x64xf32>)
      outs(%0:tensor<16x64xf32>)
      dimensions = [1]
      (%in: f32, %out: f32) {
        %1 = arith.addf %out, %in: f32
        linalg.yield %1: f32
      }
  %2 = tensor.expand_shape %reduce [[0],[1, 2]] : tensor<16x64xf32> into tensor<16x1x64xf32>
```

In the vector dialect, it will be:
```
    %cst = arith.constant 0.000000e+00 : f32
    %c0 = arith.constant 0 : index
    %0 = tensor.empty() : tensor<16x64xf32>
    %1 = vector.transfer_read %arg0[%c0, %c0, %c0], %cst {in_bounds = [true, true, true]} : tensor<16x32x64xf32>, vector<16x32x64xf32>
    %2 = vector.transfer_read %0[%c0, %c0], %cst {in_bounds = [true, true]} : tensor<16x64xf32>, vector<16x64xf32>
    %3 = vector.multi_reduction <add>, %1, %2 [1] : vector<16x32x64xf32> to vector<16x64xf32>
    %4 = vector.shape_cast %3 : vector<16x64xf32> to vector<16x1x64xf32>
    %5 = tensor.empty() : tensor<16x1x64xf32>
    %6 = vector.transfer_write %4, %5[%c0, %c0, %c0] {in_bounds = [true, true, true]} : vector<16x1x64xf32>, tensor<16x1x64xf32>
```

Reply:
No, linalg.reduce will not accept same rank after reduce, we will use linalg.reduce + tensor.expand_shape.